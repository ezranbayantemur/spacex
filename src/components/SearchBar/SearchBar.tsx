import React from 'react';
import {View} from 'react-native';
import {useFuseSearch} from '@hooks';
import Input from '../Input';
import type {SearchBarProps} from './SearchBar.types';

const SearchBar = <T,>({
  data,
  testID,
  searchKeys,
  placeholder = 'Search...',
  debounceTime = 200,
  onSearch,
}: SearchBarProps<T>) => {
  const {result, search} = useFuseSearch<T>(data, {
    keys: searchKeys,
    threshold: 0.1,
  });

  const handleOnSearch = (text: string) => {
    search(text);
  };

  React.useEffect(() => {
    if (!result || !onSearch) {
      return;
    }

    onSearch(result);
  }, [result, onSearch]);

  return (
    <View>
      <Input
        testID={`${testID}_searchbar`}
        placeholder={placeholder}
        debounceTime={debounceTime}
        onChangeText={handleOnSearch}
        autoCapitalize="none"
      />
    </View>
  );
};

export default React.memo(SearchBar);
