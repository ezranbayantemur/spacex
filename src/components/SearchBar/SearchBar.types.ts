export interface SearchBarProps<T> {
  testID: string;
  data: T[];
  searchKeys: string[];
  onSearch: (result: T[]) => void;
  placeholder?: string;
  debounceTime?: number;
}
