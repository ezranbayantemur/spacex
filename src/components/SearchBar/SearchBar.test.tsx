import React from 'react';
import {act, fireEvent, render} from '@testing-library/react-native';
import launches from '@mocks/launches.json';
import SearchBar from './SearchBar';

const mockOnSearch = jest.fn();

let wrapper: ReturnType<typeof render>;

jest.useFakeTimers();
describe('SearchBar unit tests', () => {
  beforeEach(() => {
    wrapper = render(
      <SearchBar
        testID="test"
        data={launches}
        searchKeys={['name']}
        onSearch={mockOnSearch}
      />,
    );
  });

  it('should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should render default placeholder correctly', () => {
    expect(wrapper.queryByPlaceholderText('Search...')).not.toBeNull();
  });

  it('should render custom placeholder correctly', () => {
    wrapper.rerender(
      <SearchBar
        data={launches}
        testID="test"
        searchKeys={['name']}
        onSearch={mockOnSearch}
        placeholder="Test placeholder..."
      />,
    );
    expect(wrapper.queryByPlaceholderText('Test placeholder...')).not.toBeNull();
  });

  it('should trigger search correctly', () => {
    const input = wrapper.getByTestId('test_searchbar_input');
    act(() => {
      fireEvent(input, 'onChangeText', 'trail');
      jest.advanceTimersByTime(200);
    });

    expect(mockOnSearch).toHaveBeenCalledWith([launches[2]]);
  });

  it('should trigger search with custom debounce correctly', () => {
    wrapper.rerender(
      <SearchBar
        testID="test"
        data={launches}
        searchKeys={['name']}
        onSearch={mockOnSearch}
        debounceTime={500}
      />,
    );

    const input = wrapper.getByTestId('test_searchbar_input');
    act(() => {
      fireEvent(input, 'onChangeText', 'trail');
      jest.advanceTimersByTime(500);
    });

    expect(mockOnSearch).toHaveBeenCalledWith([launches[2]]);
  });
});
