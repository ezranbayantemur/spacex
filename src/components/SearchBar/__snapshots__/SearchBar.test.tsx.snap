// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBar unit tests should render correctly 1`] = `
<View>
  <View>
    <View
      style={
        {
          "borderBottomColor": "#E1E1E1",
          "borderBottomWidth": 1,
          "margin": 16,
          "marginVertical": 8,
          "paddingBottom": 8,
          "paddingTop": 16,
        }
      }
    >
      <TextInput
        autoCapitalize="none"
        onChangeText={[Function]}
        placeholder="Search..."
        style={
          {
            "fontFamily": "Quicksand-Regular",
            "fontSize": 15,
            "padding": 0,
          }
        }
        testID="test_searchbar_input"
      />
    </View>
  </View>
</View>
`;
